mixnet/packet.py

```diff
@@ -0,0 +1,203 @@
+from __future__ import annotations
+
+import uuid
+from dataclasses import dataclass
+from typing import Dict, List, Self, Tuple, TypeAlias
+
+from mixnet.mixnet import Mixnet, MixnetTopology, MixNode
+from mixnet.sphinx.payload import Payload
+from mixnet.sphinx.sphinx import SphinxPacket
+
+MessageFlag: TypeAlias = bytes  # 1byte
+
+MESSAGE_FLAG_REAL: MessageFlag = b"\x00"
+MESSAGE_FLAG_DROP_COVER: MessageFlag = b"\x01"
+
+
+class PacketBuilder:
+    """Build a real packet or a drop cover packet."""
+
+    @staticmethod
+    def build_real_packets(
+        message: bytes, mixnet: Mixnet, topology: MixnetTopology
+    ) -> Tuple[List[SphinxPacket], List[List[MixNode]]]:
+        """Build real SphinxPackets and return them with a list of mix routes chosen for each SphinxPacket respectively."""
+        destination = mixnet.choose_mixnode()
+
+        msg_with_flag = MESSAGE_FLAG_REAL + message
+        # NOTE: We don't encrypt msg_with_flag for destination.
+        # If encryption is needed, a shared secret must be appended in front of the message along with the MessageFlag.
+        fragment_set = FragmentSet(msg_with_flag)
+
+        packets = []
+        routes = []
+        for fragment in fragment_set.fragments:
+            route = topology.generate_route()
+            packets.append(SphinxPacket.build(fragment.bytes(), route, destination))
+            routes.append(route)
+        return packets, routes
+
+    @staticmethod
+    def build_drop_cover_packet(
+        mixnet: Mixnet, topology: MixnetTopology
+    ) -> Tuple[SphinxPacket, List[MixNode]]:
+        """Bulid a drop cover SphinxPacket and return it with a mix route chosen"""
+        destination = mixnet.choose_mixnode()
+
+        msg_with_flag = MESSAGE_FLAG_DROP_COVER + b""
+        # NOTE: We don't encrypt msg_with_flag for destination.
+        # If encryption is needed, a shared secret must be appended in front of the message along with the MessageFlag.
+        fragment_set = FragmentSet(msg_with_flag)
+        # Since a drop cover message is very small (1byte + padding),
+        # it should fit in a single fragment.
+        assert len(fragment_set.fragments) == 1
+
+        route = topology.generate_route()
+        return (
+            SphinxPacket.build(fragment_set.fragments[0].bytes(), route, destination),
+            route,
+        )
```

Comment:
Good point. I'll add abstraction. 

> Also, I'm curious why the drop_cover is just a single packet while the real one is set (shouldn't it be wrapped already?)

A `message` for real traffic could be bigger than the capacity of a single packet (fragment). In that case, the `message` is split into multiple packets. That's why the `build_real_packets` returns a list. 
A cover traffic doesn't need to be large. It's enough to generate a single cover packet whenever needed, according to the emission rate `lambda` defined in https://www.notion.so/Mixnet-Specification-807b624444a54a4b88afa1cc80e100c2#f0b5080b9e764d02b69729a3f8567b63.

Reply:
Ah, I think we don't need to differentiate real and cover. A user can decide how long the message for cover would be. Of course, it's recommended to set it to fit into a single packet to meet the expected emission rate easily.

I refactored this `PacketBuilder` instead of defining two classes that inherit `SphinxPacket`. I think this is cleaner. Please have a look when you have time. 91c8e1e01814ca53c9853c0f6ab1852d76ce44a3

Reply:
> Good point. I'll add abstraction.
> 
> > Also, I'm curious why the drop_cover is just a single packet while the real one is set (shouldn't it be wrapped already?)
> 
> A `message` for real traffic could be bigger than the capacity of a single packet (fragment). In that case, the `message` is split into multiple packets. That's why the `build_real_packets` returns a list. A cover traffic doesn't need to be large. It's enough to generate a single cover packet whenever needed, according to the emission rate `lambda` defined in https://www.notion.so/Mixnet-Specification-807b624444a54a4b88afa1cc80e100c2#f0b5080b9e764d02b69729a3f8567b63.

Thanks for the explanation!!